cargo/cargo_build_script.bzl

```diff
@@ -79,8 +69,9 @@ def _build_script_impl(ctx):
     env = dict(ctx.configuration.default_shell_env)
 
     env.update({
+        "CARGO_CRATE_NAME": name_to_crate_name(pkg_name),
```

Comment:
Nit: `name_to_crate_name` is documented as taking a Bazel target name, but here we're applying it to a Cargo package name. True, it does the right thing, but we distinguish between Bazel target names and Cargo package names in other places (e.g. in the `_name_to_pkg_name` function you're introducing in this change), so using `name_to_crate_name` in this way muddies the distinction a bit.

Here are two alternatives:
- Eliminate `name_to_crate_name` entirely and replace it with `.replace("-", "_")`. (Note that `x.replace("-", "_")` actually has _fewer_ characters than `name_to_crate_name(x)`. For this reason, I would also not suggest simply renaming `name_to_crate_name` to something more generic, such as `hyphens_to_underscores`, because that would say exactly the same thing as `.replace("-", "_")`, just with more characters.)
- If `name_to_crate_name` _is_ a useful function, then it's not because it saves characters (see above) but because it introduces a useful abstraction: It converts one type of name to another without us having to be aware of the rules for doing so. If we believe this abstraction is useful (I do), it should be crisp about the type of name it accepts and returns, so we should provide differently named functions, e.g. `target_name_to_crate_name` and `pkg_name_to_crate_name`, even if they happen to have the same implementation. (Note that `cargo_build_script.bzl` contains another use of `name_to_crate_name` to convert a Cargo feature name to the name of the corresponding `CARGO_FEATURE_` environment variable. I plan to replace this usage with a function that is specific to feature names.)
  If you go for this alternative, I realize it does mean some churn. I don't think all existing uses of `name_to_crate_name` would need to be replaced in this PR -- that could (and probably should) happen in a separate PR. Happy to help out with this too.

Reply:
I can see the value in being specific about what kind of name the function accepts, but in this particular case I'm not so sure... In general `CARGO_PKG_NAME` and `ctx.label.name` are identical, but for build scripts we may optionally strip a suffix, which is a fairly minor transformation.

If we're going to change something here, I'd lean towards renaming `_name_to_pkg_name` to something like `maybe_strip_build_script_suffix`, and continue treating `ctx.label.name` and `CARGO_PKG_NAME` as identical concepts (so maybe renaming `name_to_crate_name` to either `pkg_name_to_crate_name` or `bazel_name_to_crate_name`, but still calling it directly here despite having technically preprocessed).

I can get on board with inlining the `CARGO_FEATURE` one to just a `.replace("-", "_")`, that feels like it's not the same underlying abstraction being replaced there and doesn't really need a named function extracted :)

WDYT?

Reply:
I don't feel too strongly about this -- it's just something that triggered my  "mental type checker", so I thought I'd point it out. What you're proposing sounds fine to me!